controller/kubernetes_pod_controller.go

```diff
@@ -0,0 +1,271 @@
+	defer kc.queue.ShutDown()
+
+	logrus.Infof("Start " + controllerAgentName)
+	if !controller.WaitForCacheSync(controllerAgentName, stopCh, kc.pStoreSynced, kc.pvStoreSynced, kc.pvcStoreSynced) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(kc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	logrus.Infof("Shutting down" + controllerAgentName)
+}
+
+func (kc *KubernetesPodController) worker() {
+	for kc.processNextWorkItem() {
+	}
+}
+
+func (kc *KubernetesPodController) processNextWorkItem() bool {
+	key, quit := kc.queue.Get()
+	if quit {
+		return false
+	}
+	defer kc.queue.Done(key)
+	err := kc.syncHandler(key.(string))
+	kc.handleErr(err, key)
+	return true
+}
+
+func (kc *KubernetesPodController) handleErr(err error, key interface{}) {
+	if err == nil {
+		kc.queue.Forget(key)
+		return
+	}
+
+	if kc.queue.NumRequeues(key) < maxRetries {
+		logrus.Warnf("Error syncing Longhorn kubernetes pod %v: %v", key, err)
+		kc.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logrus.Warnf("Dropping Longhorn kubernetes pod %v out of the queue: %v", key, err)
+	kc.queue.Forget(key)
+}
+
+func (kc *KubernetesPodController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, controllerAgentName+": fail to sync %v", key)
+	}()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	pod, err := kc.pLister.Pods(namespace).Get(name)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "Error getting Pod: %s", name)
+	}
+	nodeID := pod.Spec.NodeName
+	kc.forceDeletePod(pod, nodeID, namespace)
+
+	return nil
+}
+
+// Force delete a pod when all of the below conditions are meet:
+// 1. StatefulSetPodDeletePolicy is StatefulSetPodDeletePolicyForceDelete
+// 2. pod is belonging to a StatefulSet
+// 3. node containing the pod is downed
+// 4. the pod is terminating
+// 5. pod has a PV with provisioner driver.longhorn.io
+func (kc *KubernetesPodController) forceDeletePod(pod *v1.Pod, nodeID string, namespace string) {
+	// by default we don't force delete pod of StatefulSet
+	var deletionPolicy types.StatefulSetPodDeletePolicy
+	if deletionSetting, err := kc.ds.GetSettingValueExisted(types.SettingNameStatefulSetPodDeletePolicy); err != nil {
+		deletionPolicy = types.StatefulSetPodDeletePolicyNever
+	} else {
+		deletionPolicy = types.StatefulSetPodDeletePolicy(deletionSetting)
+	}
+
+	if deletionPolicy == types.StatefulSetPodDeletePolicyNever {
+		return
+	}
+
+	if !isOwnedByStatefulSet(pod) {
+		return
+	}
+
+	// only delete StatefulSet pods on the node which is declared `NotReady` or doesn't exist.
+	isNodeDowned, err := kc.ds.IsNodeDownOrDeleted(nodeID)
+	if err != nil {
+		logrus.Errorf("failed to evaluate Node %v for pod %v in forceDeletePod: %v", nodeID, pod.Name, err)
+		return
+	}
+	if !isNodeDowned {
+		return
+	}
+
+	if pod.DeletionTimestamp != nil && pod.DeletionTimestamp.After(time.Now()) {
```

Comment:
Consider waiting till the `deletionTimestamp` has passed, since that is the point at which the kubelet on the node should have deleted the pod, but since it didn't you can confidently assume that the node/kubelet is in a state where it's no longer able to communicate with the cluster.

Reply:
Does it `pod.DeletionTimestamp.After(time.Now())` makes sure the `deletionTimestamp` has passed?

Reply:
see [comment](https://github.com/longhorn/longhorn-manager/pull/640#discussion_r465347701)